### PR TITLE
Mace update

### PR DESCRIPTION
**Pull request by Hanwen1018**
update mace potential to the newest version

### CODE REVIEW

File: mlptrain/potentials/mace/mace.py [hunk 1]
```diff
@@ -388,442 +85,140 @@ def valid_fraction(self) -> float:
...
+            str(Config.mace_params['config_type_weights']),
+            '--E0s',
+            str(self.get_E0s),
+            '--model',
+            str(Config.mace_params['model']),
+            '--hidden_irreps',
+            str(Config.mace_params['hidden_irreps']),
+            '--r_max',
+            str(Config.mace_params['r_max']),
+            '--lr',
```
Comment (juraskov) [inline mlptrain/potentials/mace/mace.py:122]: What is lr, and why it is not tunable?

File: mlptrain/potentials/mace/mace.py [hunk 2]
```diff
@@ -388,442 +85,140 @@ def valid_fraction(self) -> float:
...
+            str(0.001),
+            '--scaling',
```
Comment (juraskov) [inline mlptrain/potentials/mace/mace.py:124]: Are there other options for scaling available?

File: mlptrain/config.py [hunk 3]
```diff
@@ -55,6 +55,7 @@ class _ConfigClass:
 
     mace_params = {
         'valid_fraction': 0.1,
+        'max_num_epochs': 1200,
         'config_type_weights': '{"Default":1.0}',
         'model': 'MACE',
```
Comment (juraskov) [inline mlptrain/config.py:60]: Could you list in comment other options for model, please?

File: mlptrain/potentials/mace/mace.py [hunk 4]
```diff
@@ -388,442 +85,140 @@ def valid_fraction(self) -> float:
...
-            'energy': True,
-            'forces': True,
-            'virials': False,
-            'stress': False,
-            'dipoles': False,
-        }
+        args_list = [
+            '--name',
+            self.name,
+            '--max_L',
```
Comment (juraskov) [inline mlptrain/potentials/mace/mace.py:102]: What is L? Should we allow user to modify it?

File: mlptrain/potentials/mace/mace.py [hunk 5]
```diff
@@ -388,442 +85,140 @@ def valid_fraction(self) -> float:
...
+            str(Config.mace_params['device']),
+            '--default_dtype',
+            str(Config.mace_params['dtype']),
+            '--seed',
+            str(345),
+            '--energy_key',
+            'energy',
+            '--forces_key',
+            'forces',
+            '--num_workers',
```
Comment (juraskov) [inline mlptrain/potentials/mace/mace.py:159]: What is num_workers?

File: mlptrain/potentials/mace/mace.py [hunk 6]
```diff
@@ -388,442 +85,140 @@ def valid_fraction(self) -> float:
...
+            str(35),
+        ]
+
+        if self.foundation is not None:
+            args_list.append('--foundation_model')
+            args_list.append(f'{self.foundation}')
+
+        if Config.mace_params['save_cpu'] is True:
```
Comment (juraskov) [inline mlptrain/potentials/mace/mace.py:167]: It seems that save_cpu is already in the list

File: mlptrain/potentials/mace/mace.py [hunk 7]
```diff
@@ -48,6 +33,8 @@ def __init__(
 
             system: (mlptrain.System) Object defining the system without
                                       specifying the coordinates
+
+            foundation: (str) Name of the foundation model used in fine-tunning
```
Comment (juraskov) [inline mlptrain/potentials/mace/mace.py:37]: Could you list here the options available?

**Review by juraskov — CHANGES_REQUESTED** [review]
Thanks for the update! Here is the first round of the things I spotted.

File: mlptrain/config.py (hunk 3 above)
Comment (Hanwen1018) [reply mlptrain/config.py:60]: you mean model selection? I think here we would only do MACE

File: mlptrain/potentials/mace/mace.py [hunk 8]
```diff
@@ -1,42 +1,27 @@
...
-    import torch_ema
-    from e3nn import o3
-    from torch.optim.swa_utils import SWALR, AveragedModel
-    from mace import data, modules, tools
-    from mace.calculators import MACECalculator
-    from mace.tools import torch_geometric
-    from mace.tools.scripts_utils import create_error_table
-except ModuleNotFoundError:
-    pass
+from mace.calculators import MACECalculator
```
[the remark this answers is not in the document]
Comment (Hanwen1018) [reply mlptrain/potentials/mace/mace.py]: added

File: mlptrain/potentials/mace/mace.py (hunk 7 above)
Comment (Hanwen1018) [reply mlptrain/potentials/mace/mace.py:41]: added

File: mlptrain/potentials/mace/mace.py [hunk 9]
```diff
@@ -388,442 +85,140 @@ def valid_fraction(self) -> float:
...
-                )
-
-                self._loss_fn = modules.EnergyForcesLoss(
-                    energy_weight=Config.mace_params['energy_weight'],
-                    forces_weight=Config.mace_params['forces_weight'],
-                )
-
-        return self._loss_fn
-
-    @property
```
[the remark this answers is not in the document]
Comment (Hanwen1018) [reply mlptrain/potentials/mace/mace.py:475]: set as "batch_size". not specific to train_batch_size or valid_batch_size

File: mlptrain/potentials/mace/mace.py [hunk 10]
```diff
@@ -388,442 +85,140 @@ def valid_fraction(self) -> float:
...
-    def train_batch_size(self) -> int:
-        """Batch size of the training set"""
-        if len(self.train_configs) < Config.mace_params['batch_size']:
-            return len(self.train_configs)
-        else:
-            return Config.mace_params['batch_size']
-
-    @property
-    def valid_batch_size(self) -> int:
```
[the remark this answers is not in the document]
Comment (Hanwen1018) [reply mlptrain/potentials/mace/mace.py:484]: Same as above

File: mlptrain/potentials/mace/mace.py (hunk 1 above)
Comment (Hanwen1018) [reply mlptrain/potentials/mace/mace.py:127]: learning rate. Better to not change as changing it will cause failure in learning like loss function not converge.

File: mlptrain/potentials/mace/mace.py (hunk 5 above)
Comment (Hanwen1018) [reply mlptrain/potentials/mace/mace.py:164]: number of workers to load the data

File: mlptrain/potentials/mace/mace.py (hunk 4 above)
Comment (Hanwen1018) [reply mlptrain/potentials/mace/mace.py:107]: equivariance layer. our GPUs can not train MACE with L=2

File: mlptrain/potentials/mace/mace.py (hunk 2 above)
Comment (Hanwen1018) [reply mlptrain/potentials/mace/mace.py:129]: other options are not recommend.

File: environment_mace.yml [hunk 11]
```diff
@@ -28,4 +28,4 @@ dependencies:
   - git
   - pip:
     - ase@git+https://gitlab.com/ase/ase.git@f2615a6e9a # For PLUMED
-    - mace-torch==0.3.6
+    - mace-torch
```
[the remark this answers is not in the document]
Comment (Hanwen1018) [reply environment_mace.yml]: added version of 0.3.10. I will do test on 0.3.12, there is no large changes from 0.3.10 to 0.3.12.

File: mlptrain/potentials/mace/mace.py (hunk 7 above)
Comment (juraskov) [reply mlptrain/potentials/mace/mace.py:41]: Btw. what type of fine tuning is default in this approach? Multihead or naive? Can you make a note on it as well?

**Review by juraskov — CHANGES_REQUESTED** [review]
Thanks Hanwen, I added some minor comments.

File: mlptrain/potentials/mace/mace.py (hunk 5 above)
Comment (Hanwen1018) [reply mlptrain/potentials/mace/mace.py:171]: added in config

File: mlptrain/potentials/mace/mace.py [hunk 12]
```diff
@@ -388,442 +92,147 @@ def valid_fraction(self) -> float:
...
+            '--E0s',
+            str(self.get_E0s),
+            '--model',
+            str(Config.mace_params['model']),
+            '--hidden_irreps',
+            str(Config.mace_params['hidden_irreps']),
+            '--r_max',
+            str(Config.mace_params['r_max']),
+            '--lr',
+            str(0.001),
```
Comment (JoeHolownia) [inline mlptrain/potentials/mace/mace.py:138]:
```suggestion
            str(Config.mace_params['lr']),
```

File: mlptrain/potentials/mace/mace.py [hunk 13]
```diff
@@ -388,442 +92,147 @@ def valid_fraction(self) -> float:
...
+            '--swa',
+            '--ema',
+            '--ema_decay',
+            str(0.999),
+            '--device',
+            str(Config.mace_params['device']),
+            '--default_dtype',
+            str(Config.mace_params['dtype']),
+            '--seed',
+            str(345),
```
Comment (JoeHolownia) [inline mlptrain/potentials/mace/mace.py:168]:
```suggestion
           str(Config.mace_params['seed']),
```

File: mlptrain/potentials/mace/mace.py [hunk 14]
```diff
@@ -388,442 +92,147 @@ def valid_fraction(self) -> float:
...
+            '--max_num_epochs',
+            str(Config.mace_params['max_num_epochs']),
+            '--error_table',
+            str(Config.mace_params['error_table']),
+            '--loss',
+            str(Config.mace_params['loss']),
+            '--correlation',
+            str(Config.mace_params['correlation']),
+            '--scheduler_patience',
+            str(20),
```
Comment (JoeHolownia) [inline mlptrain/potentials/mace/mace.py:154]:
```suggestion
           str(Config.mace_params['scheduler_patience']),
```

File: mlptrain/potentials/mace/mace.py [hunk 15]
```diff
@@ -388,442 +92,147 @@ def valid_fraction(self) -> float:
...
+            '--patience',
+            str(50),
```
Comment (JoeHolownia) [inline mlptrain/potentials/mace/mace.py:156]:
```suggestion
            str(Config.mace_params['patience']),
```

File: mlptrain/potentials/mace/mace.py
Comment (JoeHolownia) [inline mlptrain/potentials/mace/mace.py:1]: I think we should be consistent in having the mace_params in config.py controlling the parameters passed to the MACE script, and put all our default values there (also enabling users to simply modify the mace_params dict). I have just made suggestions to move the default values all to the mace_params in config.py

File: mlptrain/potentials/mace/mace.py [hunk 16]
```diff
@@ -388,442 +92,147 @@ def valid_fraction(self) -> float:
...
+            '--start_swa',
+            str(Config.mace_params['start_swa']),
+            '--swa',
+            '--ema',
```
Comment (JoeHolownia) [inline mlptrain/potentials/mace/mace.py:160]: The ema and swa options in config.py are ignored here, so let's check them and optionally add the ema parameters like we do for 'foundation_model' etc. below

File: mlptrain/potentials/mace/mace.py [hunk 17]
```diff
@@ -388,442 +92,147 @@ def valid_fraction(self) -> float:
...
+        ]
+
+        if self.foundation is not None:
+            args_list.append('--foundation_model')
+            args_list.append(f'{self.foundation}')
+
+        if Config.mace_params['save_cpu'] is True:
+            args_list.append('--save_cpu')
+
+        if Config.mace_params['restart_latest'] is True:
```
Comment (JoeHolownia) [inline mlptrain/potentials/mace/mace.py:184]: is True for a boolean is redundant (except for readability I suppose...)
```suggestion
        if Config.mace_params['restart_latest']:
```

File: mlptrain/potentials/mace/mace.py [hunk 18]
```diff
@@ -388,442 +92,147 @@ def valid_fraction(self) -> float:
...
+            'forces',
+            '--num_workers',
+            str(Config.mace_params['num_workers']),
+        ]
+
+        if self.foundation is not None:
+            args_list.append('--foundation_model')
+            args_list.append(f'{self.foundation}')
+
+        if Config.mace_params['save_cpu'] is True:
```
Comment (JoeHolownia) [inline mlptrain/potentials/mace/mace.py:181]: is True for a boolean is redundant (except for readability I suppose...)
```suggestion
        if Config.mace_params['save_cpu']:
```

File: mlptrain/potentials/mace/mace.py [hunk 19]
```diff
@@ -388,442 +92,147 @@ def valid_fraction(self) -> float:
...
+            '--seed',
+            str(345),
+            '--energy_key',
+            'energy',
+            '--forces_key',
+            'forces',
+            '--num_workers',
+            str(Config.mace_params['num_workers']),
+        ]
+
```
Comment (JoeHolownia) [inline mlptrain/potentials/mace/mace.py:176]: As mentioned above, lets use the config.py mace_params to optionally add these parameters.
```suggestion

        if Config.mace_params['ema']:
                args_list.append('--ema')
                args_list.append('--ema_decay')
                args_list.append(str(Config.mace_params['ema_decay']))
      
        if Config.mace_params['swa']:
                args_list.append('--swa')
                args_list.append('--start_swa')
                args_list.append(str(Config.mace_params['start_swa']))
```

File: mlptrain/config.py [hunk 20]
```diff
@@ -74,6 +75,8 @@ class _ConfigClass:
         'amsgrad': True,
```
Comment (JoeHolownia) [inline mlptrain/config.py:75]: Add patience and scheduler_patience control.
```suggestion
        'lr': 0.001,
        'patience': 50,
        'scheduler_patience': 20,
        'amsgrad': True,
```

**Review by JoeHolownia — CHANGES_REQUESTED** [review]
I've just suggested we move the defaults of the MACE script calling into config.py mace_params for consistency. Some of the parameters in mace_params (such as ema / swa, are ignored currently).
